anvil_consortium_manager/models.py

```diff
@@ -79,6 +82,30 @@ def anvil_import(cls, billing_project_name):
         billing_project.save()
         return billing_project
 
+    @classmethod
+    def anvil_audit(cls):
+        """Verify data in the app against AnVIL.
+
+        Only billing projects with have_app_as_user=True are checked, because the AnVIL API does not
+        differentiate between billing projects that don't exist and billing projects where the app is
+        not a user.
+
+        Raises:
+            NotFoundInAnVIL: if any billing projects were not found in AnVIL.
+
+        Returns:
+            None if there are no issues."""
+        # Check that all billing projects exist.
+        missing = []
+        for billing_project in cls.objects.filter(has_app_as_user=True).all():
+            if not billing_project.anvil_exists():
+                missing.append("{} {}".format(billing_project.pk, billing_project.name))
+        if missing:
+            raise cls.DoesNotExistInAnVIL(
+                "Not a user of billing projects on AnVIL: {}".format(",".join(missing))
+            )
+        # Here we don't care if there are billing projects in AnVIL that we haven't imported into the app.
+
```

Comment:
Should this return `True`? Currently it returns `None` if there were no errors and raises an exception if there were any issues.

Reply:
Instead of raising an exception, maybe it should return a dictionary of errors.

Reply:
Alternatively, make a class to hold audit results instead of using a dictionary.

Reply:
I'm definitely starting to prefer the class over a dictionary of lists for a couple reasons:
- the class could define a "truthiness" or "success" function to indicate if the audit was successful or if there were any errors.
- a class has a more defined structure, compared to a list of dictionaries generated on the fly.
- the class could have a method to create and/or send an automated email with the audit results.

Reply:
Switched to using a class: see classes in `anvil_audit.py`.